File: days/day_08/solutions/solution.ex
defmodule Solution do
  defmodule Coord do
    defstruct [:x, :y, :z]
  end

  defmodule Pair do
    defstruct [:i, :j, :dist_sq]
  end

  def parse_coordinates(lines) do
    lines
    |> Enum.map(&String.trim/1)
    |> Enum.filter(&(&1 != ""))
    |> Enum.map(fn line ->
      [x, y, z] = line |> String.split(",") |> Enum.map(&String.to_integer/1)
      %Coord{x: x, y: y, z: z}
    end)
  end

  def squared_distance(%Coord{} = p1, %Coord{} = p2) do
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    dz = p2.z - p1.z
    dx * dx + dy * dy + dz * dz
  end

  def solve(input_data) do
    lines = input_data |> String.trim() |> String.split("\n")
    coords = parse_coordinates(lines)
    n = length(coords)

    if n == 0 do
      {"0", "0"}
    else
      # Generate all pairs with squared distances
      pairs =
        if n > 1 do
          0..(n - 1)
          |> Enum.flat_map(fn i ->
            if i + 1 <= n - 1 do
              for j <- (i + 1)..(n - 1) do
                dist_sq = squared_distance(Enum.at(coords, i), Enum.at(coords, j))
                %Pair{i: i, j: j, dist_sq: dist_sq}
              end
            else
              []
            end
          end)
        else
          []
        end
        |> Enum.sort_by(& &1.dist_sq)

      # Part 1: Connect first 1000 pairs
      uf1 = init_union_find(n)
      {uf1, _} =
        pairs
        |> Enum.take(1000)
        |> Enum.reduce({uf1, 0}, fn pair, {uf, conn_count} ->
          {union(uf, pair.i, pair.j), conn_count + 1}
        end)

      # Get component sizes
      component_sizes =
        if n > 0 do
          0..(n - 1)
          |> Enum.reduce(%{}, fn i, acc ->
            root = find(uf1, i)
            Map.update(acc, root, get_size(uf1, root), fn _ -> get_size(uf1, root) end)
          end)
        else
          %{}
        end

      sizes = component_sizes |> Map.values() |> Enum.sort(:desc)
      part1 = if length(sizes) >= 3 do
        [s1, s2, s3 | _] = sizes
        s1 * s2 * s3
      else
        0
      end

      # Part 2: Connect until all in one circuit
      uf2 = init_union_find(n)
      final_pair =
        pairs
        |> Enum.reduce_while({uf2, nil}, fn pair, {uf, _} ->
          if uf.component_count == 1 do
            {:halt, {uf, nil}}
          else
            new_uf = union(uf, pair.i, pair.j)
            if new_uf.component_count == 1 do
              {:halt, {new_uf, pair}}
            else
              {:cont, {new_uf, nil}}
            end
          end
        end)
        |> elem(1)

      part2 = if final_pair do
        coord_i = Enum.at(coords, final_pair.i)
        coord_j = Enum.at(coords, final_pair.j)
        coord_i.x * coord_j.x
      else
        0
      end

      {Integer.to_string(part1), Integer.to_string(part2)}
    end
  end

  def init_union_find(n) do
    %{
      parent: for(i <- 0..(n - 1), do: i) |> List.to_tuple(),
      size: for(_ <- 1..n, do: 1) |> List.to_tuple(),
      component_count: n
    }
  end

  def find(uf, x) do
    p = elem(uf.parent, x)
    if p == x do
      x
    else
      root = find(uf, p)
      # Path compression not shown here for simplicity (would need mutable state)
      root
    end
  end

  def get_size(uf, root) do
    elem(uf.size, root)
  end

  def union(uf, x, y) do
    root_x = find(uf, x)
    root_y = find(uf, y)
    if root_x == root_y do
      uf
    else
      size_x = get_size(uf, root_x)
      size_y = get_size(uf, root_y)
      {rx, ry} = if size_x < size_y, do: {root_y, root_x}, else: {root_x, root_y}
      new_parent = uf.parent |> Tuple.to_list() |> List.replace_at(ry, rx) |> List.to_tuple()
      new_size = uf.size |> Tuple.to_list() |> List.replace_at(rx, get_size(uf, rx) + get_size(uf, ry)) |> List.to_tuple()
      %{uf | parent: new_parent, size: new_size, component_count: uf.component_count - 1}
    end
  end
end

{part1, part2} =
  "../data/input.txt"
  |> File.read!()
  |> Solution.solve()

IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")

File: days/day_08/solutions/solution.hs
import Data.List hiding (union)
import Data.Maybe
import qualified Data.Map.Strict as Map

data Coord = Coord { x :: Integer, y :: Integer, z :: Integer } deriving (Eq, Show)

type Pair = (Int, Int, Integer)

data UnionFind = UF {
    parent :: [Int],
    size :: [Integer],
    componentCount :: Int
} deriving (Show)

parseCoordinates :: [String] -> [Coord]
parseCoordinates lines = map parseLine (filter (not . null . trim) lines)
  where
    trim = dropWhile (== ' ')
    parseLine line =
        let parts = split ',' (trim line)
            x = read (parts !! 0) :: Integer
            y = read (parts !! 1) :: Integer
            z = read (parts !! 2) :: Integer
        in Coord x y z
    split _ [] = []
    split c s = let (part, rest) = break (== c) s
                in part : case rest of
                    [] -> []
                    (_:rs) -> split c rs

squaredDistance :: Coord -> Coord -> Integer
squaredDistance p1 p2 =
    let dx = x p2 - x p1
        dy = y p2 - y p1
        dz = z p2 - z p1
    in dx * dx + dy * dy + dz * dz

initUnionFind :: Int -> UnionFind
initUnionFind n = UF {
    parent = [0..n-1],
    size = replicate n 1,
    componentCount = n
}

findRoot :: UnionFind -> Int -> (UnionFind, Int)
findRoot uf x =
    let p = parent uf !! x
    in if p == x
       then (uf, x)
       else let (uf', root) = findRoot uf p
                newParent = take x (parent uf') ++ [root] ++ drop (x + 1) (parent uf')
            in (uf' { parent = newParent }, root)

union :: UnionFind -> Int -> Int -> UnionFind
union uf x y =
    let (uf1, rootX) = findRoot uf x
        (uf2, rootY) = findRoot uf1 y
    in if rootX == rootY
       then uf2
       else let sizeX = size uf2 !! rootX
                sizeY = size uf2 !! rootY
                (rx, ry) = if sizeX < sizeY then (rootY, rootX) else (rootX, rootY)
                newParent = take ry (parent uf2) ++ [rx] ++ drop (ry + 1) (parent uf2)
                newSize = take rx (size uf2) ++ [size uf2 !! rx + size uf2 !! ry] ++ drop (rx + 1) (size uf2)
            in uf2 { parent = newParent, size = newSize, componentCount = componentCount uf2 - 1 }

solve :: String -> (String, String)
solve inputData =
    let lines' = lines inputData
        coords = parseCoordinates lines'
        n = length coords
    in if n == 0
       then ("0", "0")
       else let -- Generate all pairs
                pairs = [(i, j, squaredDistance (coords !! i) (coords !! j)) | i <- [0..n-1], j <- [i+1..n-1]]
                sortedPairs = sortOn (\(_, _, dist) -> dist) pairs
                
                -- Part 1: Connect first 1000 pairs
                uf1 = foldl (\uf (i, j, _) -> union uf i j) (initUnionFind n) (take 1000 sortedPairs)
                
                -- Get component sizes
                componentSizes = Map.fromList $ map (\i -> let (_, root) = findRoot uf1 i in (root, size uf1 !! root)) [0..n-1]
                sizes = sortBy (flip compare) $ Map.elems componentSizes
                part1 = if length sizes >= 3 then product (take 3 sizes) else 0
                
                -- Part 2: Connect until all in one circuit
                finalPair = findFinalPair (initUnionFind n) sortedPairs
                part2 = case finalPair of
                    Just (i, j, _) -> x (coords !! i) * x (coords !! j)
                    Nothing -> 0
            in (show part1, show part2)

findFinalPair :: UnionFind -> [Pair] -> Maybe Pair
findFinalPair uf [] = Nothing
findFinalPair uf ((i, j, dist):pairs)
    | componentCount uf == 1 = Nothing
    | otherwise = let newUF = union uf i j
                  in if componentCount newUF == 1
                     then Just (i, j, dist)
                     else findFinalPair newUF pairs

main :: IO ()
main = do
    input <- readFile "../data/input.txt"
    let (part1, part2) = solve input
    putStrLn $ "Part 1: " ++ part1
    putStrLn $ "Part 2: " ++ part2
